Properly Restore Gas Lift Optimisation Parameters at Restart

bska opened the pull request:
In particular

  * Treat negative restart file GLO limit values as "no limit".
  * Apply proper unit conversion for the "minimum economic gradient threshold".  The `identity` conversion is not applicable to FIELD units.
  * Load GLO parameters early when constructing groups and wells.  This fixes an unintended consequence of PR #2794 (commit 1849e223) which would lose any GLO parameters for groups directly parented to FIELD.

Together these fixes enable restarting the [`4_GLIFT_MODEL5`](https://github.com/OPM/opm-tests/blob/7d8c76583622448c6f78ecb2587c304ac83cd0ed/model5/4_GLIFT_MODEL5.DATA) example model from OPM-Tests and having the restarted run at least respect **an** upper limit on the total lift gas injection rate where the current master sources do not respect any such limit.  See below for an example.  However, we do yet not meet the prescribed [200,000 SM3/DAY](https://github.com/OPM/opm-tests/blob/7d8c76583622448c6f78ecb2587c304ac83cd0ed/model5/4_GLIFT_MODEL5.DATA#L323) limit, but rather that 200,000 SM3/DAY limit minus the [12,500 SM3/DAY increment](https://github.com/OPM/opm-tests/blob/7d8c76583622448c6f78ecb2587c304ac83cd0ed/model5/4_GLIFT_MODEL5.DATA#L318) which suggests that there are other issues still remaining here.

I nevertheless submit this work for review to at least somewhat improve the current situation with respect to restarting runs with group level gas lift optimisation.

---

#### Example Restarted Run of `4_GLIFT_MODEL5` ####

We compare the group level gas lift injection rate (`GGLIR`) for the `PLAT-A` group parented directly to the FIELD group.  The current master sources (`*_RST_MASTER`) do not honour any upper bound on the gas lift injection rate, whereas the current PR (`*_RST_DEV`) honours an incorrect, lower, limit on this gas lift injection rate.

![PLAT_A_Gas_Lift_Injection_Rate_master_dev_compare](https://github.com/user-attachments/assets/0592efc1-a697-4946-8a8f-4389a415ecff)

bska commented:
jenkins build this please

akva2 commented on opm/input/eclipse/Schedule/GasLiftOpt.cpp:
```diff
@@ -39,6 +67,34 @@ bool GasLiftGroup::operator==(const GasLiftGroup& other) const
+// found in the restart file is pure guesswork.
+bool GasLiftWell::active(const RestartIO::RstWell& rst_well)
+{
+    if (rst_well.glift_max_rate +
```
could be return a+b+c != 0

bska replied:
> could be return a+b+c != 0

Good point.  I've pushed an update to that effect.  It's not unlikely, though, that we'll revise this condition in the future.

akva2 commented on opm/input/eclipse/Schedule/GasLiftOpt.cpp:
```diff
@@ -17,12 +17,40 @@
 
+bool GasLiftGroup::active(const RestartIO::RstGroup& rst_group)
+{
+    if (rst_group.glift_max_rate +
```
could be return a+b != 0

bska replied:
> could be return a+b != 0

Good point.  I've pushed an update to that effect.  It's not unlikely, though, that we'll revise this condition in the future.

akva2 commented on opm/input/eclipse/Schedule/GasLiftOpt.hpp:
```diff
@@ -181,43 +311,165 @@ class GasLiftWell {
+    /// \param[in] gaslift_increment Lift gas rate increment value.
     void gaslift_increment(double gaslift_increment);
+
+    /// Retreive minimum economical gradient threshold to continue
```
Retrieve

bska replied:
> Retrieve

Good catch.  I've pushed an update to address this.

akva2 reviewed (approved):
Just some small stuff

bska commented:
Thank you for the review.  I believe I have addressed all comments in the latest update here and I'll run a build check.  For what it's worth, I ran an experiment where I increased the maximum lift gas injection rate of the `PLAT-A` group from [200,000 SM3/DAY](https://github.com/OPM/opm-tests/blob/7d8c76583622448c6f78ecb2587c304ac83cd0ed/model5/4_GLIFT_MODEL5.DATA#L323) to 200,000.5 SM3/DAY while leaving all other parameters unchanged. In that case, the restarted run&ndash;curve below&ndash;does reach the 200,000 SM3/DAY level of the base run although we still do not quite reproduce the actual behaviour of the base run.

---

![Slack_PLAT_A_Gas_Lift_Injection_Rate](https://github.com/user-attachments/assets/51b26a98-f269-4d57-b11e-d6db8a8c3955)

bska commented:
jenkins build this please
